enclave/core/sgx/calls.c

```diff
@@ -1246,9 +1233,54 @@ void oe_abort(void)
         __oe_enclave_status = OE_ENCLAVE_ABORTING;
     }
 
-    // Free the shared memory pools
-    oe_teardown_arena();
+    /* Abort can be called with user-modified FS (e.g., FS check fails in
+     * oe_sgx_get_td()). To avoid using is_enclave_debug_allowed(), which
+     * depends on oe_sgx_get_td(), we base only on the cached value to
+     * determine if the enclave is in debug mode. If the
+     * is_enclave_debug_allowed has not beed called before, the following
+     * function will return false and the stack stitching will be skipped. */
+    if (is_enclave_debug_allowed_cached())
+    {
+        oe_ecall_context_t* host_ecall_context = td->host_ecall_context;
+
+        // Make sure the context is valid.
+        if (host_ecall_context &&
+            oe_is_outside_enclave(
+                host_ecall_context, sizeof(*host_ecall_context)))
+        {
+            uint64_t* frame = (uint64_t*)__builtin_frame_address(0);
+            host_ecall_context->debug_eexit_rbp = frame[0];
+            // The caller's RSP is always given by this equation
+            //   RBP + 8 (caller frame pointer) + 8 (caller return address)
+            host_ecall_context->debug_eexit_rsp = frame[0] + 8;
+            host_ecall_context->debug_eexit_rip = frame[1];
+        }
+    }
 
     // Return to the latest ECALL.
-    _handle_exit(OE_CODE_ERET, 0, __oe_enclave_status);
+    oe_asm_exit(arg1, __oe_enclave_status, td, 1 /* direct_return */);
+}
+
+void oe_abort(void)
+{
+    /* Bypass the FS check given that the oe_abort can be invoked anywhere */
+    oe_sgx_td_t* td = oe_sgx_get_td_no_fs_check();
+
+    /* It is unlikely that td is invalid. If this is the case, we cannot
+     * call _abort to exit the enclave. Instead, we intentionally trigger
```

Comment:
Nice idea to immediately exit the enclave via page fault. 
Is there a way to also get other threads to immediately exit the enclave when one thread aborts?

Reply:
This may require some signal-like mechanisms. We can think about it later.